source/common/router/router.cc

```diff
@@ -844,11 +844,33 @@ void Filter::cleanup() {
   }
 }
 
+std::string Filter::getShadowCluster(const ShadowPolicy& policy,
+                                     const Http::HeaderMap& headers) const {
+  if (!policy.cluster().empty()) {
+    return policy.cluster();
+  } else {
+    ASSERT(!policy.clusterHeader().get().empty());
+    const auto entry = headers.get(policy.clusterHeader());
+    if (!entry.empty()) {
+      return std::string(entry[0]->value().getStringView());
+    }
+    ENVOY_STREAM_LOG(warn, "There is no cluster name in header: {}", *callbacks_,
+                     policy.clusterHeader());
+    return "";
+  }
+}
+
 void Filter::maybeDoShadowing() {
   for (const auto& shadow_policy_wrapper : active_shadow_policies_) {
     const auto& shadow_policy = shadow_policy_wrapper.get();
 
-    ASSERT(!shadow_policy.cluster().empty());
+    const std::string cluster = getShadowCluster(shadow_policy, *downstream_headers_);
+
+    // The cluster name in cluster_header is empty
```

Comment:
nit: end comments with '.'.

Now that the PGV constraint is lifted, couldn't it mean that `cluster` is set to an empty string?
(I'm guessing not, because of the validation, but want to make sure)

Reply:
Exact one of `cluster` and `cluster_header` should be not empty from the validation. Both a policy config with `cluster: "", cluster_header: some_header` and one with only `cluster_header:some_header` can pass the runtime config validation. In the runtime config validation, we only check if the header name exists. If the `cluster` in config is not empty, we will use that value. Otherwise, we need to get the exact cluster name from the headers, eg: `some_header: cluster_1`. It is possible that there is no this header or the header value is empty. In that case, we returned `absl::nullopt`. 

Then we skip the remaining steps.